File: clients/net-chat-go/internal/bitchat/protocol/message_type.go
// Package protocol — Go port of bitchat-rust's src/protocol/*.rs.
//
// Wire-compatible with mainline permissionlesstech/bitchat (iOS + Android)
// and the bigsony fork. Kept pure — no BLE, no allocation beyond what the
// packet itself needs, no external deps beyond golang.org/x/crypto for the
// crypto layer above.
package protocol

import "fmt"

// MessageType is the 1-byte type tag in a packet header. Numeric values
// are mainline-compatible — see message_type.rs comment on 0x02 vs 0x04.
type MessageType uint8

const (
	MTAnnounce                 MessageType = 0x01
	MTMessage                  MessageType = 0x02
	MTLeave                    MessageType = 0x03
	MTFragmentStart            MessageType = 0x05
	MTFragmentContinue         MessageType = 0x06
	MTFragmentEnd              MessageType = 0x07
	MTChannelAnnounce          MessageType = 0x08
	MTChannelRetention         MessageType = 0x09
	MTDeliveryAck              MessageType = 0x0A
	MTDeliveryStatusRequest    MessageType = 0x0B
	MTReadReceipt              MessageType = 0x0C
	MTNoiseHandshakeInit       MessageType = 0x10
	MTNoiseHandshakeResp       MessageType = 0x11
	MTNoiseEncrypted           MessageType = 0x12
	MTNoiseIdentityAnnounce    MessageType = 0x13
	MTChannelKeyVerifyRequest  MessageType = 0x14
	MTChannelKeyVerifyResponse MessageType = 0x15
	MTChannelPasswordUpdate    MessageType = 0x16
	MTChannelMetadata          MessageType = 0x17
	MTNoiseHandshakeFinal      MessageType = 0x18
	MTVersionHello             MessageType = 0x20
	MTVersionAck               MessageType = 0x21
	MTProtocolAck              MessageType = 0x22
	MTProtocolNack             MessageType = 0x23
	MTSystemValidation         MessageType = 0x24
	MTHandshakeRequest         MessageType = 0x25
	MTFavorited                MessageType = 0x30
	MTUnfavorited              MessageType = 0x31
)

// MessageTypeFromByte parses a wire byte. Legacy 0x04 aliases to
// MTMessage — mirrors message_type.rs's rolling-upgrade tolerance.
func MessageTypeFromByte(b byte) (MessageType, error) {
	switch b {
	case 0x01:
		return MTAnnounce, nil
	case 0x02, 0x04:
		return MTMessage, nil
	case 0x03:
		return MTLeave, nil
	case 0x05:
		return MTFragmentStart, nil
	case 0x06:
		return MTFragmentContinue, nil
	case 0x07:
		return MTFragmentEnd, nil
	case 0x08:
		return MTChannelAnnounce, nil
	case 0x09:
		return MTChannelRetention, nil
	case 0x0A:
		return MTDeliveryAck, nil
	case 0x0B:
		return MTDeliveryStatusRequest, nil
	case 0x0C:
		return MTReadReceipt, nil
	case 0x10:
		return MTNoiseHandshakeInit, nil
	case 0x11:
		return MTNoiseHandshakeResp, nil
	case 0x12:
		return MTNoiseEncrypted, nil
	case 0x13:
		return MTNoiseIdentityAnnounce, nil
	case 0x14:
		return MTChannelKeyVerifyRequest, nil
	case 0x15:
		return MTChannelKeyVerifyResponse, nil
	case 0x16:
		return MTChannelPasswordUpdate, nil
	case 0x17:
		return MTChannelMetadata, nil
	case 0x18:
		return MTNoiseHandshakeFinal, nil
	case 0x20:
		return MTVersionHello, nil
	case 0x21:
		return MTVersionAck, nil
	case 0x22:
		return MTProtocolAck, nil
	case 0x23:
		return MTProtocolNack, nil
	case 0x24:
		return MTSystemValidation, nil
	case 0x25:
		return MTHandshakeRequest, nil
	case 0x30:
		return MTFavorited, nil
	case 0x31:
		return MTUnfavorited, nil
	}
	return 0, fmt.Errorf("unknown message type 0x%02x", b)
}

File: clients/net-chat-android/app/src/main/java/io/grfx/netchat/protocol/MessageType.java
package io.grfx.netchat.protocol;

/**
 * BitChat message type byte. Numeric values MUST match mainline
 * permissionlesstech/bitchat + the Rust fork so we interop with
 * iOS/Android clients out of the box. See bitchat-rust
 * src/protocol/message_type.rs for the same table.
 *
 * <p>Note the 0x02 / 0x04 dual mapping for Message: mainline uses 0x02,
 * the older fork used 0x04. From-byte accepts either → Message.
 */
public enum MessageType {
    ANNOUNCE(0x01),
    MESSAGE(0x02),
    LEAVE(0x03),
    FRAGMENT_START(0x05),
    FRAGMENT_CONTINUE(0x06),
    FRAGMENT_END(0x07),
    CHANNEL_ANNOUNCE(0x08),
    CHANNEL_RETENTION(0x09),
    DELIVERY_ACK(0x0A),
    DELIVERY_STATUS_REQUEST(0x0B),
    READ_RECEIPT(0x0C),
    NOISE_HANDSHAKE_INIT(0x10),
    NOISE_HANDSHAKE_RESP(0x11),
    NOISE_ENCRYPTED(0x12),
    NOISE_IDENTITY_ANNOUNCE(0x13),
    CHANNEL_KEY_VERIFY_REQUEST(0x14),
    CHANNEL_KEY_VERIFY_RESPONSE(0x15),
    CHANNEL_PASSWORD_UPDATE(0x16),
    CHANNEL_METADATA(0x17),
    NOISE_HANDSHAKE_FINAL(0x18),
    VERSION_HELLO(0x20),
    VERSION_ACK(0x21),
    PROTOCOL_ACK(0x22),
    PROTOCOL_NACK(0x23),
    SYSTEM_VALIDATION(0x24),
    HANDSHAKE_REQUEST(0x25),
    FAVORITED(0x30),
    UNFAVORITED(0x31);

    public final byte value;

    MessageType(int v) { this.value = (byte) v; }

    /**
     * Decode a wire byte to a MessageType. Legacy 0x04 aliases to
     * MESSAGE so a rolling upgrade across the fork doesn't strand
     * older nodes.
     *
     * @throws UnknownMessageTypeException on any value not in the table.
     */
    public static MessageType fromByte(byte b) throws UnknownMessageTypeException {
        int i = b & 0xFF;
        switch (i) {
            case 0x01: return ANNOUNCE;
            case 0x02: case 0x04: return MESSAGE;
            case 0x03: return LEAVE;
            case 0x05: return FRAGMENT_START;
            case 0x06: return FRAGMENT_CONTINUE;
            case 0x07: return FRAGMENT_END;
            case 0x08: return CHANNEL_ANNOUNCE;
            case 0x09: return CHANNEL_RETENTION;
            case 0x0A: return DELIVERY_ACK;
            case 0x0B: return DELIVERY_STATUS_REQUEST;
            case 0x0C: return READ_RECEIPT;
            case 0x10: return NOISE_HANDSHAKE_INIT;
            case 0x11: return NOISE_HANDSHAKE_RESP;
            case 0x12: return NOISE_ENCRYPTED;
            case 0x13: return NOISE_IDENTITY_ANNOUNCE;
            case 0x14: return CHANNEL_KEY_VERIFY_REQUEST;
            case 0x15: return CHANNEL_KEY_VERIFY_RESPONSE;
            case 0x16: return CHANNEL_PASSWORD_UPDATE;
            case 0x17: return CHANNEL_METADATA;
            case 0x18: return NOISE_HANDSHAKE_FINAL;
            case 0x20: return VERSION_HELLO;
            case 0x21: return VERSION_ACK;
            case 0x22: return PROTOCOL_ACK;
            case 0x23: return PROTOCOL_NACK;
            case 0x24: return SYSTEM_VALIDATION;
            case 0x25: return HANDSHAKE_REQUEST;
            case 0x30: return FAVORITED;
            case 0x31: return UNFAVORITED;
        }
        throw new UnknownMessageTypeException(b);
    }

    public static final class UnknownMessageTypeException extends Exception {
        public final byte value;
        UnknownMessageTypeException(byte v) {
            super(String.format("unknown message type 0x%02x", v & 0xFF));
            this.value = v;
        }
    }
}
